File: backup/auto-backup-2025-09-12-1858/src/lib/firebase/repositories/serviceRepository.ts
import { 
  collection,
  doc,
  getDocs,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc,
  query,
  where,
  orderBy,
  onSnapshot,
  Timestamp,
  Unsubscribe
} from 'firebase/firestore';
import { db } from '../config';
import { Service } from '@/types/models';

/**
 * Firebase Service Repository - Replicates Android service queries
 * Handles services with same patterns as Android app
 */
class FirebaseServiceRepository {
  private static instance: FirebaseServiceRepository;
  
  private constructor() {}
  
  static getInstance(): FirebaseServiceRepository {
    if (!FirebaseServiceRepository.instance) {
      FirebaseServiceRepository.instance = new FirebaseServiceRepository();
    }
    return FirebaseServiceRepository.instance;
  }
  
  /**
   * Get all services for a specific shop (matches Android pattern)
   * Uses combined query to satisfy Firestore security rules
   */
  async getServicesByShop(shopId: string): Promise<Service[]> {
    try {
      console.log(`🔥 Fetching services for shop ${shopId}`);
      
      // Combined query to satisfy Firestore security rules (only active services can be read)
      const servicesQuery = query(
        collection(db, 'services'),
        where('shopId', '==', shopId),
        where('active', '==', true)
      );
      
      const snapshot = await getDocs(servicesQuery);
      console.log(`🔥 Fetched ${snapshot.docs.length} active service documents for shop`);
      
      const services: Service[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const service: Service = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Service;
          
          console.log(`✅ Successfully parsed active service for shop: ${service.name}`);
          services.push(service);
        } catch (e) {
          console.error(`❌ Failed to parse service document ${document.id}`, e);
          // Skip invalid documents like Android does
        }
      });
      
      // Sort in code instead of query (like Android does)
      return services.sort((a, b) => b.createdAt - a.createdAt);
    } catch (error) {
      console.error('❌ Error fetching shop services:', error);
      return [];
    }
  }
  
  /**
   * Get all active services (for marketplace view)
   */
  async getActiveServices(): Promise<Service[]> {
    try {
      console.log('🔥 Fetching all active services');
      
      // Simplified query without orderBy to avoid index requirements
      const servicesQuery = query(
        collection(db, 'services'),
        where('active', '==', true)
      );
      
      const snapshot = await getDocs(servicesQuery);
      console.log(`🔥 Fetched ${snapshot.docs.length} active service documents`);
      
      const services: Service[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const service: Service = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Service;
          
          console.log(`✅ Successfully parsed active service: ${service.name}`);
          services.push(service);
        } catch (e) {
          console.error(`❌ Failed to parse service document ${document.id}`, e);
        }
      });
      
      console.log(`✅ Successfully parsed ${services.length} active services`);
      return services;
    } catch (error) {
      console.error('❌ Error fetching active services:', error);
      return [];
    }
  }
  
  /**
   * Get a service by its unique ID
   */
  async getServiceById(id: string): Promise<Service | null> {
    try {
      const docRef = doc(db, 'services', id);
      const document = await getDoc(docRef);
      
      if (!document.exists()) {
        return null;
      }
      
      const data = document.data();
      const service: Service = {
        id: document.id,
        ...data,
        createdAt: data.createdAt || Date.now(),
        updatedAt: data.updatedAt || Date.now(),
      } as Service;
      
      return service;
    } catch (error) {
      console.error(`❌ Error fetching service ${id}:`, error);
      return null;
    }
  }
  
  /**
   * Get services by owner ID (for business dashboard)
   */
  async getServicesByOwner(ownerId: string): Promise<Service[]> {
    try {
      const servicesQuery = query(
        collection(db, 'services'),
        where('ownerId', '==', ownerId),
        orderBy('createdAt', 'desc')
      );
      
      const snapshot = await getDocs(servicesQuery);
      
      const services: Service[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const service: Service = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Service;
          
          services.push(service);
        } catch (e) {
          console.error(`❌ Failed to parse service document ${document.id}`, e);
        }
      });
      
      return services;
    } catch (error) {
      console.error(`❌ Error fetching services for owner ${ownerId}:`, error);
      return [];
    }
  }
  
  /**
   * Get bookable services for a shop (only services where isBookable = true)
   */
  async getBookableServicesByShop(shopId: string): Promise<Service[]> {
    try {
      const servicesQuery = query(
        collection(db, 'services'),
        where('shopId', '==', shopId),
        where('isActive', '==', true),
        where('isBookable', '==', true)
      );
      
      const snapshot = await getDocs(servicesQuery);
      
      const services: Service[] = [];
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const service: Service = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Service;
          
          services.push(service);
        } catch (e) {
          console.error(`❌ Failed to parse bookable service document ${document.id}`, e);
        }
      });
      
      return services.sort((a, b) => b.createdAt - a.createdAt);
    } catch (error) {
      console.error(`❌ Error fetching bookable services for shop ${shopId}:`, error);
      return [];
    }
  }
  
  /**
   * Add a new service
   */
  async addService(service: Omit<Service, 'id'>): Promise<string | null> {
    try {
      const serviceData = {
        ...service,
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      
      const docRef = await addDoc(collection(db, 'services'), serviceData);
      return docRef.id;
    } catch (error) {
      console.error('❌ Error adding service:', error);
      return null;
    }
  }
  
  /**
   * Update an existing service
   */
  async updateService(service: Service): Promise<boolean> {
    try {
      const docRef = doc(db, 'services', service.id);
      const { id, createdAt, ...updateData } = service;
      
      await updateDoc(docRef, {
        ...updateData,
        updatedAt: Date.now(),
      });
      
      return true;
    } catch (error) {
      console.error(`❌ Error updating service ${service.id}:`, error);
      return false;
    }
  }
  
  /**
   * Delete a service
   */
  async deleteService(id: string): Promise<boolean> {
    try {
      const docRef = doc(db, 'services', id);
      await deleteDoc(docRef);
      return true;
    } catch (error) {
      console.error(`❌ Error deleting service ${id}:`, error);
      return false;
    }
  }
  
  /**
   * Search services by name or description
   */
  async searchServices(searchTerm: string, limit: number = 20): Promise<Service[]> {
    try {
      const servicesQuery = query(
        collection(db, 'services'),
        where('isActive', '==', true),
        orderBy('name')
      );
      
      const snapshot = await getDocs(servicesQuery);
      
      const services: Service[] = [];
      const searchLower = searchTerm.toLowerCase();
      
      snapshot.docs.forEach((document) => {
        try {
          const data = document.data();
          const service: Service = {
            id: document.id,
            ...data,
            createdAt: data.createdAt || Date.now(),
            updatedAt: data.updatedAt || Date.now(),
          } as Service;
          
          // Simple text matching
          if (
            service.name.toLowerCase().includes(searchLower) ||
            service.description.toLowerCase().includes(searchLower) ||
            service.tags.some(tag => tag.toLowerCase().includes(searchLower))
          ) {
            services.push(service);
          }
        } catch (e) {
          console.error(`❌ Failed to parse service document ${document.id}`, e);
        }
      });
      
      return services.slice(0, limit);
    } catch (error) {
      console.error('❌ Error searching services:', error);
      return [];
    }
  }
}

// Export singleton instance
export const serviceRepository = FirebaseServiceRepository.getInstance();

File: androidAPP/MyHustle/app/src/main/java/com/blueclipse/myhustle/data/repository/ServiceRepository.kt
package com.blueclipse.myhustle.data.repository

import com.blueclipse.myhustle.data.model.Service
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/**
 * Repository for managing service data in Firebase
 */
class ServiceRepository private constructor() {
    private val firestore = FirebaseFirestore.getInstance()
    private val servicesCollection = firestore.collection("services")
    
    private val _services = MutableStateFlow<List<Service>>(emptyList())
    val services: StateFlow<List<Service>> = _services.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    companion object {
        @Volatile
        private var INSTANCE: ServiceRepository? = null
        
        val instance: ServiceRepository
            get() = INSTANCE ?: synchronized(this) {
                INSTANCE ?: ServiceRepository().also { INSTANCE = it }
            }
    }
    
    /**
     * Create a new service
     */
    suspend fun createService(service: Service): Result<String> {
        return try {
            _isLoading.value = true
            val documentRef = servicesCollection.add(service).await()
            val serviceId = documentRef.id
            
            // Update the service with its ID
            documentRef.update("id", serviceId).await()
            
            Result.success(serviceId)
        } catch (e: Exception) {
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get services for a specific shop
     */
    suspend fun getServicesForShop(shopId: String): Result<List<Service>> {
        return try {
            _isLoading.value = true
            val snapshot = servicesCollection
                .whereEqualTo("shopId", shopId)
                .whereEqualTo("isActive", true)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .get()
                .await()
            
            val services = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Service::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }
            
            _services.value = services
            Result.success(services)
        } catch (e: Exception) {
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get all active services
     */
    suspend fun getAllServices(): Result<List<Service>> {
        return try {
            _isLoading.value = true
            val snapshot = servicesCollection
                .whereEqualTo("isActive", true)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .get()
                .await()
            
            val services = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Service::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }
            
            Result.success(services)
        } catch (e: Exception) {
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    /**
     * Get service by ID
     */
    suspend fun getServiceById(serviceId: String): Result<Service?> {
        return try {
            val document = servicesCollection.document(serviceId).get().await()
            if (document.exists()) {
                val service = document.toObject(Service::class.java)?.copy(id = document.id)
                Result.success(service)
            } else {
                Result.success(null)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update service
     */
    suspend fun updateService(service: Service): Result<Unit> {
        return try {
            val updateData = service.copy(updatedAt = System.currentTimeMillis())
            servicesCollection.document(service.id).set(updateData).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Delete service
     */
    suspend fun deleteService(serviceId: String): Result<Unit> {
        return try {
            servicesCollection.document(serviceId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Get services by category
     */
    suspend fun getServicesByCategory(category: String): Result<List<Service>> {
        return try {
            val snapshot = servicesCollection
                .whereEqualTo("category", category)
                .whereEqualTo("isActive", true)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .get()
                .await()
            
            val services = snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Service::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }
            
            Result.success(services)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
